### PR TITLE
Rewrite TINC tests for AO checksum failure, and move to main test suite.

### PR DESCRIPTION
**Pull request by hlinnaka**
Some notable differences:

* For the last "appendonly_verify_block_checksums_co" test, don't use a
  compressed table. That seems fragile.

* In the "appendonly_verify_block_checksums_co" test, be more explicit in
  what is replaced. It used to scan backwards from end of file for the byte
  'a', but we didn't explicitly include that byte in the test data. What
  actually gets replaced depends heavily on how integers are encoded. (And
  the table was compressed too, which made it even more indeterministic.)
  In the rewritten test, we replace the string 'xyz', and we use a text
  field that contains that string as the table data.

* Don't restore the original table file after corrupting. That seemed
  uninteresting to test. Presumably the table was OK before we corrupted
  it, so surely it's still OK after restoring it back. In theory, there
  could be a problem if the file's corrupt contents were cached somewhere,
  but we don't cache AO tables, and I'm not sure what we'd try to prove by
  testing that, anway, because swapping the file while the system is active
  is surely not supported.

* The old script checked that the output when a corrupt table was SELECTed
  from contained the string "ERROR:  Header checksum does not match".
  However, half of the tests actually printed a different error, "Block
  checksum does not match". It turns out that the way the old select_table
  function asserted the result of the grep command was wrong. It should've
  done "assert(int(result) > 0), ..." rather than just "assert(result > 0),
  ...". As written, it always passed, even if there was no ERROR in the
  output. The rewritten test does not have that bug.

### CODE REVIEW

File: src/test/regress/expected/ao_checksum_corruption.out [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+   0
+(1 row)
+
+SELECT COUNT(*) FROM corrupt_content_small_ao;
+ERROR:  Block checksum does not match.  Expected 0x30FEE3C5 and found 0xA6847419 (cdbappendonlystorageread.c:1184)  (seg0 slice1 127.0.0.1:40000 pid=28137)
+-- Also test gp_appendonly_verify_block_checksums=off.
+create table appendonly_verify_block_checksums_co (t text)  with (checksum=true, appendonly=true, orientation=column, compresstype=none);
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 't' as the Greenplum Database data distribution key for this table.
+insert into appendonly_verify_block_checksums_co
+  select 'abcdefghijlmnopqrstuvxyz' from generate_series(1, 50);
```
Comment (ashwinstar) [inline src/test/regress/expected/ao_checksum_corruption.out]: Should we cut down number of rows ?

File: src/test/regress/expected/ao_checksum_corruption.out [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+NOTICE:  append only table corrupt_checksums.corrupt_content_small_ao depends on schema corrupt_checksums
+NOTICE:  append only table corrupt_checksums.corrupt_header_small_ao depends on schema corrupt_checksums
+NOTICE:  append only columnar table corrupt_checksums.corrupt_content_small_co depends on schema corrupt_checksums
+NOTICE:  append only columnar table corrupt_checksums.corrupt_header_small_co depends on schema corrupt_checksums
+NOTICE:  append only columnar table corrupt_checksums.corrupt_content_large_co depends on schema corrupt_checksums
+NOTICE:  append only columnar table corrupt_checksums.corrupt_header_large_co depends on schema corrupt_checksums
+NOTICE:  function corrupt_checksums.corrupt_file(text,bytea,bytea) depends on schema corrupt_checksums
+NOTICE:  function corrupt_checksums.corrupt_file(text,integer) depends on schema corrupt_checksums
+NOTICE:  function corrupt_checksums.get_aoseg1_path(regclass) depends on schema corrupt_checksums
+ERROR:  cannot drop schema corrupt_checksums because other objects depend on it
```
Comment (ashwinstar) [inline src/test/regress/expected/ao_checksum_corruption.out]: Seems the schema is not dropped.

**Comment by ivannovick** [comment]
Thx @hlinnaka for this!

**Comment by hlinnaka** [comment]
On 03/12/2017 07:38 AM, Ashwin Agrawal wrote:
> ashwinstar approved this pull request.
>
> Amazing 😲 ! Really you just keep amazing me with how much can be done and that too so simply in ICW 🙇

Technically, you can hook up anything to installcheck-world. It's just a 
makefile target, it can run whatever :-)

>> +select SUM(corrupt_file(get_aoseg1_path('corrupt_content_small_ao'), -3)) from gp_dist_random('gp_id');
> +INFO:  corrupting file base/107062/107111.1 at -3  (seg0 slice1 127.0.0.1:40000 pid=28137)
> +INFO:  corrupting file base/107062/107111.1 at -3  (seg1 slice1 127.0.0.1:40001 pid=28138)
> +INFO:  corrupting file base/107062/107111.1 at -3  (seg2 slice1 127.0.0.1:40002 pid=28139)
> + sum
> +-----
> +   0
> +(1 row)
> +
> +SELECT COUNT(*) FROM corrupt_content_small_ao;
> +ERROR:  Block checksum does not match.  Expected 0x30FEE3C5 and found 0xA6847419 (cdbappendonlystorageread.c:1184)  (seg0 slice1 127.0.0.1:40000 pid=28137)
> +-- Also test gp_appendonly_verify_block_checksums=off.
> +create table appendonly_verify_block_checksums_co (t text)  with (checksum=true, appendonly=true, orientation=column, compresstype=none);
> +NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 't' as the Greenplum Database data distribution key for this table.
> +insert into appendonly_verify_block_checksums_co
> +  select 'abcdefghijlmnopqrstuvxyz' from generate_series(1, 50);
>
> Should we cut down number of rows ?

I actually already cut it down from 100 in the original test to 50. I 
wasn't 100% sure if the size makes some difference to the AO table's 
block structure. But it's not totally clear to me what the point of this 
test was in the first place, so yeah, I guess we might as well cut it 
down further.

> +-- Clean up. We don't want to leave the corrupt tables lying around!
> +reset search_path;
> +DROP SCHEMA corrupt_checksums;
> +NOTICE:  append only columnar table corrupt_checksums.appendonly_verify_block_checksums_co depends on schema corrupt_checksums
> +NOTICE:  append only table corrupt_checksums.corrupt_content_small_ao depends on schema corrupt_checksums
> +NOTICE:  append only table corrupt_checksums.corrupt_header_small_ao depends on schema corrupt_checksums
> +NOTICE:  append only columnar table corrupt_checksums.corrupt_content_small_co depends on schema corrupt_checksums
> +NOTICE:  append only columnar table corrupt_checksums.corrupt_header_small_co depends on schema corrupt_checksums
> +NOTICE:  append only columnar table corrupt_checksums.corrupt_content_large_co depends on schema corrupt_checksums
> +NOTICE:  append only columnar table corrupt_checksums.corrupt_header_large_co depends on schema corrupt_checksums
> +NOTICE:  function corrupt_checksums.corrupt_file(text,bytea,bytea) depends on schema corrupt_checksums
> +NOTICE:  function corrupt_checksums.corrupt_file(text,integer) depends on schema corrupt_checksums
> +NOTICE:  function corrupt_checksums.get_aoseg1_path(regclass) depends on schema corrupt_checksums
> +ERROR:  cannot drop schema corrupt_checksums because other objects depend on it
>
> Seems the schema is not dropped.

Ouch, how sloppy of me. Will fix, thanks!

- Heikki

**Comment by hlinnaka** [comment]
Pushed, after adding CASCADE to the DROP at the end, and reducing the number of rows in that test from 50 to 5. Thanks for the review!

**Pull request closed without merging**